### ISSUE TITLE
PointConv "add_self_loops" causes gradients to flow between elements in a batch

### ISSUE THREAD

**Issue by beneisner** [rusty1s/pytorch_geometric#2558]
## 🐛 Bug

When running batch training on any model that uses [PointConv](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.PointConv) (i.e. PointNet++, as provided in the examples folder), enabling "add_self_loops" on PointConv causes, for some reason, causes gradients to get mixed up between examples in the batch.

Concretely, if we have a pointcloud batch (x) with two pointclouds such that x = [x_1, x_2], and a neural network f that produces outputs y = [y_1, y_2] = f(x) and uses PointConv under the hood (i.e. PointNet++), then we would expect that the following Jacobians should always be observed:

First, we have expected gradients, where input element corresponds to output element.
d(y_1)/d(x_1) := probably nonzero, this is the expected gradient.
d(y_2)/d(x_2) := probably nonzero, this is the expected gradient.

When output element doesn't correspond to input elements, we should expect Jacobians with 0 elements everywhere!
Intuitively this makes sense, outputs in a batch should only be a function of their inputs.
d(y_1)/d(x_2) = 0 
d(y_2)/d(x_1) = 0

This assumes that we have "batch norm" or other cross-batch things turned off, or in "eval" mode, which should remove that as a source of cross-batch gradients.

HOWEVER! This invariant is violated when PointConv is called with `add_self_loops=True` in the constructor, and called in the fashion demonstrated by the example code for Pointnet++. My understanding is that this option should enable each point embedding to be conditioned on its neighborhood as well as itself. This makes sense, and I think is consistent with the original paper.

When `add_self_loops=True`, the misaligned Jacobians (i.e. output doesn't correspond to input) are often nonzero. Below, I've modified the example code found in the [Pointnet++ Classification example](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/pointnet2_classification.py) to illustrate the behavior.

My only changes were to:
1. add a `reproduce` function that constructs a batch and computes the corresponding gradients
2. Pipe through the `add_self_loops` flag so that it can be controlled from `main`

## To Reproduce

1. Run the code below, and watch the assert fail.
2. Change `WITH_SELF_LOOPS = False` in the `main` block and rerun, and see that the assert passes.

```
import os.path as osp

import torch
import torch.nn.functional as F
from torch.nn import Sequential as Seq, Linear as Lin, ReLU, BatchNorm1d as BN
from torch_geometric.datasets import ModelNet
import torch_geometric.transforms as T
from torch_geometric.data import DataLoader, Batch, Data
from torch_geometric.nn import PointConv, fps, radius, global_max_pool


class SAModule(torch.nn.Module):
    def __init__(self, ratio, r, nn, add_self_loops):
        super(SAModule, self).__init__()
        self.ratio = ratio
        self.r = r
        self.conv = PointConv(nn, add_self_loops=add_self_loops)

    def forward(self, x, pos, batch):
        idx = fps(pos, batch, ratio=self.ratio)
        row, col = radius(pos, pos[idx], self.r, batch, batch[idx],
                          max_num_neighbors=64)
        edge_index = torch.stack([col, row], dim=0)
        x = self.conv(x, (pos, pos[idx]), edge_index)
        pos, batch = pos[idx], batch[idx]
        return x, pos, batch


class GlobalSAModule(torch.nn.Module):
    def __init__(self, nn):
        super(GlobalSAModule, self).__init__()
        self.nn = nn

    def forward(self, x, pos, batch):
        x = self.nn(torch.cat([x, pos], dim=1))
        x = global_max_pool(x, batch)
        pos = pos.new_zeros((x.size(0), 3))
        batch = torch.arange(x.size(0), device=batch.device)
        return x, pos, batch


def MLP(channels, batch_norm=True):
    return Seq(*[
        Seq(Lin(channels[i - 1], channels[i]), ReLU(), BN(channels[i]))
        for i in range(1, len(channels))
    ])


class Net(torch.nn.Module):
    def __init__(self, add_self_loops):
        super(Net, self).__init__()

        self.sa1_module = SAModule(0.5, 0.2, MLP([3, 64, 64, 128]), add_self_loops)
        self.sa2_module = SAModule(0.25, 0.4, MLP([128 + 3, 128, 128, 256]), add_self_loops)
        self.sa3_module = GlobalSAModule(MLP([256 + 3, 256, 512, 1024]))

        self.lin1 = Lin(1024, 512)
        self.lin2 = Lin(512, 256)
        self.lin3 = Lin(256, 10)

    def forward(self, data):
        sa0_out = (data.x, data.pos, data.batch)
        sa1_out = self.sa1_module(*sa0_out)
        sa2_out = self.sa2_module(*sa1_out)
        sa3_out = self.sa3_module(*sa2_out)
        x, pos, batch = sa3_out

        x = F.relu(self.lin1(x))
        x = F.dropout(x, p=0.5, training=self.training)
        x = F.relu(self.lin2(x))
        x = F.dropout(x, p=0.5, training=self.training)
        x = self.lin3(x)
        return F.log_softmax(x, dim=-1)


def train(epoch):
    model.train()

    for data in train_loader:
        data = data.to(device)
        optimizer.zero_grad()
        loss = F.nll_loss(model(data), data.y)
        loss.backward()
        optimizer.step()


def test(loader):
    model.eval()

    correct = 0
    for data in loader:
        data = data.to(device)
        with torch.no_grad():
            pred = model(data).max(1)[1]
        correct += pred.eq(data.y).sum().item()
    return correct / len(loader.dataset)

def reproduce(loader):
    # Set the model in "eval" mode, so that batch norm doesn't interfere with the analysis.
    model.eval()

    # Extract a single batch.
    batch = next(iter(loader)).to(device)
    dlist = batch.to_data_list()

    # Get two distinct training examples from the batch. We'll just look at positions.
    pos1 = dlist[0].pos
    pos2 = dlist[1].pos

    # Tell torch to collect gradient for these inputs, so we can call autograd later.
    pos1.requires_grad = True
    pos2.requires_grad = True

    # Reassemble the batch with these gradient collected inputs. There's probably a 
    # cleaner way of doing this.
    recreated_batch = Batch.from_data_list([
        Data(x=dlist[0].x, pos=pos1, y=dlist[0].y),
        Data(x=dlist[1].x, pos=pos2, y=dlist[1].y),
    ])

    # Make our predictions. This is a 32x10 matrix.
    y = model(recreated_batch)
    y1 = y[0]
    y2 = y[1]

    # If batch prediction is all correct, the partial derivatives of each batch element in the output
    # should be exactly 0 for the batch inputs that don't correspond to it.

    # Partial derivatives between predictions that correspond with their inputs. These can be anything,
    # but many elements are likely non-zero.
    dy1_dx1 = torch.autograd.grad(y1.sum(), pos1, retain_graph=True)
    dy2_dx2 = torch.autograd.grad(y2.sum(), pos2, retain_graph=True)
    
    # Partial derivatives between predictions that are mismatched with their inputs. Should be zero.
    dy1_dx2 = torch.autograd.grad(y1.sum(), pos2, retain_graph=True)
    dy2_dx1 = torch.autograd.grad(y2.sum(), pos1, retain_graph=True)

    # When WITH_SELF_LOOPS is True, the second assert fails. However, when WITH_SELF_LOOPS is False,
    # it does not fail.
    assert (dy1_dx2[0] == 0.0).all()
    assert (dy2_dx1[0] == 0.0).all()


if __name__ == '__main__':
    # CHANGE THIS FLAG TO "False", and the gradients end up being correct.
    WITH_SELF_LOOPS = True

    path = osp.join(
        osp.dirname(osp.realpath(__file__)), '..', 'data/ModelNet10')
    pre_transform, transform = T.NormalizeScale(), T.SamplePoints(1024)
    train_dataset = ModelNet(path, '10', True, transform, pre_transform)
    test_dataset = ModelNet(path, '10', False, transform, pre_transform)
    train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True,
                              num_workers=6)
    test_loader = DataLoader(test_dataset, batch_size=32, shuffle=False,
                             num_workers=6)

    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model = Net(WITH_SELF_LOOPS).to(device)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

    reproduce(train_loader)
```

## Expected behavior

See description. Both asserts in the sample code should pass - the partial derivatives between elements in the batch should be 0.

## Environment

* OS: Linux
* Python version: 3.8.6
* PyTorch version: 1.8.1
* CUDA/cuDNN version: 11.1
* GCC version: N/A 
* Any other relevant information: Evaluated on GPU.

## Additional context

I'm somewhat new to pointcloud learning, and very new to pytorch_geometric. So I'm not sure if this is intended behavior and I'm just using things wrong, or if it's an actual bug. Happy to learn either way.

Also wanted to say - I've really appreciated this library, it's made prototyping substantially easier, particularly with partial-view pointclouds that may be of different sizes. Thanks for all your hard work on the project!

**Comment by rusty1s** [comment]
Great spotting. This is indeed a bug, but it's easily fixable by adding `add_self_loops=False` to `PointConv`. I fixed it in master. Here, we do not really want to add self-loops to the graph as we are operating in bipartite graphs. The real "self-loop" is already added to `PointConv` by the `radius` call.

**Comment by beneisner** [comment]
Thanks for addressing this so quickly! Glad it was such a simple fix, I’ve updated my research code accordingly.

That said, I’m still a bit confused why add_self_loops=True causes the gradient to be contaminated? Shouldn’t all of the batch labels that are passed around keep everything straight?

**Comment by rusty1s** [comment]
It's because `add_self_loops` expects operating in a symmetric graph, i.e., it will add self loops from source node i to target node i. However, since the assignment matrix given in `edge_index` is non-symmetric (message passing to a reduced set of nodes), source node i and target node i do not necessarily denote the same node or even the same graph.
